### PR TITLE
V2 Auto Multiline Detection Foundation - Aggregator, Labeler, and Refactoring

### PR DESCRIPTION
**Pull request by gh123man**
### What does this PR do?

This PR lays the foundation for v2 auto multiline detection in the logs agent. It introduces the core concepts needed to continue developing the feature as a minimum but functional set of changes. 

The major components of this architecture form a pipeline where logs are processed by an ordered set of heuristics. Heuristics add labels or mutate the `messageContext` in order determine a label in a future step. Once a label is assigned, the aggregator aggregates the log only on the assigned label. 

Each component can be reviewed in it's own commit within this PR: 

- [Refactoring](https://github.com/DataDog/datadog-agent/pull/27881/commits/ecedeefe388b657524cee4eadb37b4c79b782709)
  - Move `TruncationFlag` and `EscapeLineFeed` into `message.go`
  - move `NewMessage` and rename to `NewRawMessage` into `message.go`
  - Line parser uses `LineHandler` interface instead of `outputFn`
  - `auto_multiline_handler` -> `legacy_auto_multiline_handler`
- [Add Labeler](https://github.com/DataDog/datadog-agent/pull/27881/commits/f6cbcccd0fc423df1cb9c6d74b3569ddb7c2b47d)
- [Add JSON detector](https://github.com/DataDog/datadog-agent/pull/27881/commits/b99fb871f3b667b15e77acfe83a831abb7865a39)
- [Add Aggregator](https://github.com/DataDog/datadog-agent/pull/27881/commits/0cbddf47ab74186e26da52589e212e7a2d52503e)
- [Add auto multiline handler + config updates](https://github.com/DataDog/datadog-agent/pull/27881/commits/eb47048b17e51a9dc6bda710336264f4a297135c)

When `logs_config.experimental_auto_multi_line_detection` is enabled - this new pipeline currently only implements one aggregation heuristic - the JSON detector which prevents structured logs from being aggregated. I chose to start with the JSON detector since it's implementation is trivial. 

More heuristic implementations and features will come in future PRs. 
This PR has no functional changes on existing features. 

### Describe how to test/QA your changes

Ive tested this locally. So `qa-done` however there isn't much to test at the moment. Complete QA steps will be added with future PRs that introduce components that flesh out the full functionality.

### CODE REVIEW

File: pkg/logs/internal/decoder/auto_multiline_detection/aggregator.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	a.bucket.add(msg)
+}
+
+func (a *Aggregator) stopFlushTimerIfNeeded() {
+	if a.flushTimer == nil || a.bucket.isEmpty() {
+		return
+	}
+	// stop the flush timer, as we now have data
+	if !a.flushTimer.Stop() {
+		<-a.flushTimer.C
```
[the remark this answers is not in the document]
Comment (gh123man) [reply pkg/logs/internal/decoder/auto_multiline_detection/aggregator.go:114]: Yep. 
It's read [here](https://github.com/DataDog/datadog-agent/blob/456dcfdc2af470855412daece4595400808f2ea6/pkg/logs/internal/decoder/decoder.go#L220). 
The whole decoder runs in a single goroutine ([introduced in this change](https://github.com/DataDog/datadog-agent/pull/11095))

File: pkg/logs/internal/decoder/auto_multiline_detection/json_detector.go [hunk 2]
```diff
@@ -8,7 +8,7 @@ package automultilinedetection
 
 import "regexp"
 
-var jsonRegexp = regexp.MustCompile(`^\s*\{\s*\"`)
+var jsonRegexp = regexp.MustCompile(`^^\s*\{\s*(\"|})`)
```
Comment (vickenty) [inline pkg/logs/internal/decoder/auto_multiline_detection/json_detector.go:11]: Duplicate `^`?

Also, may be a smidge simpler with `["}]`:

```suggestion
var jsonRegexp = regexp.MustCompile(`^\s*\{\s*["}]`)
```

**Review by vickenty — APPROVED** [review]

File: pkg/logs/internal/decoder/auto_multiline_detection/json_detector.go (hunk 2 above)
Comment (gh123man) [reply pkg/logs/internal/decoder/auto_multiline_detection/json_detector.go:11]: 🤦 . Copy paste error :( 
Surprised the go regex engine was happy with this. 
Also good idea - will do.

**Review by lukesteensen — APPROVED** [review]
Files owned by APR look good

**Review by dustmop — APPROVED** [review]
LGTM for agent-shared-components owned files

**Review by jcstorms1 — APPROVED** [review]

**Pull request merged**